lib/style-engine/class-wp-style-engine-gutenberg.php

```diff
@@ -0,0 +1,166 @@
+	 *
+	 * @since 5.5.0
+	 * @var WP_Style_Engine_Gutenberg|null
+	 */
+	private static $instance = null;
+
+	/**
+	 * Style definitions that contain the instructions to
+	 * parse/output valid Gutenberg styles from a block's attributes.
+	 * For every style definition, the follow properties are valid:
+	 *
+	 *  - property_key => the key that represents a valid CSS property, e.g., "margin" or "border".
+	 *  - value_func   => a function to generate an array of valid CSS rules for a particular style object.
+	 *                    For example, `'padding' => 'array( 'top' => '1em' )` will return `array( 'padding-top' => '1em' )`
+	 */
+	const BLOCK_STYLE_DEFINITIONS_METADATA = array(
+		'spacing' => array(
+			'padding' => array(
+				'property_key' => 'padding',
+				'path'         => array( 'spacing', 'padding' ),
+				'value_func'   => 'gutenberg_get_style_engine_css_box_rules',
+			),
+			'margin'  => array(
+				'property_key' => 'margin',
+				'path'         => array( 'spacing', 'margin' ),
+				'value_func'   => 'gutenberg_get_style_engine_css_box_rules',
+			),
+		),
+	);
+
+	/**
+	 * Utility method to retrieve the main instance of the class.
+	 *
+	 * The instance will be created if it does not exist yet.
+	 *
+	 * @return WP_Style_Engine_Gutenberg The main instance.
+	 */
+	public static function get_instance() {
+		if ( null === self::$instance ) {
+			self::$instance = new self();
+		}
+
+		return self::$instance;
+	}
+
+	/**
+	 * Returns a CSS ruleset based on the instructions in BLOCK_STYLE_DEFINITIONS_METADATA.
+	 *
+	 * @param string|array  $style_value A single raw Gutenberg style attributes value for a CSS property.
+	 * @param array<string> $path        An array of strings representing a path to the style value.
+	 *
+	 * @return array A CSS ruleset compatible with add_style().
+	 */
+	protected function get_block_style_css_rules( $style_value, $path ) {
+		$style_definition = _wp_array_get( static::BLOCK_STYLE_DEFINITIONS_METADATA, $path, null );
+
+		if ( ! empty( $style_definition ) ) {
+			if (
+				isset( $style_definition['value_func'] ) &&
+				is_callable( $style_definition['value_func'] )
+			) {
+				return call_user_func( $style_definition['value_func'], $style_value, $style_definition['property_key'] );
+			}
+		}
+
+		return array();
+	}
+
+	/**
+	 * Returns an CSS ruleset.
+	 * Styles are bundled based on the instructions in BLOCK_STYLE_DEFINITIONS_METADATA.
+	 *
+	 * @param array $block_styles An array of styles from a block's attributes.
+	 * @param array $options = array(
+	 *     'inline' => (boolean) Whether to return inline CSS rules destined to be inserted in an HTML `style` attribute.
+	 *     'path'   => (array)   Specify a block style to generate, otherwise it'll try all in BLOCK_STYLE_DEFINITIONS_METADATA.
+	 * );.
+	 *
+	 * @return string A CSS ruleset formatted to be placed in an HTML `style` attribute.
+	 */
+	public function generate( $block_styles, $options = array() ) {
+		$output = '';
```

Comment:
Not a blocker at all for now, but I was wondering if we should move this `$output` and the `return $output` into the `if` block further down that checks for the `inline` option?

In follow-ups where we add in styles that aren't generated inline, we might use a different return type that could include both the generated styles and the classnames that refer to those styles. But we can play with all that in follow-ups, I just wanted to jot down the thought 😀

Reply:
> Not a blocker at all for now, but I was wondering if we should move this $output and the return $output into the if block further down that checks for the inline option?

I declare it at the top so that the return value is always consistent. The benefit, at least for now, is that you can concatenate the return value from `generate()` to an existing string, e.g., in `spacing.php`.

Also we return `$output` early in the first if block.

Would it still be better to  move them given that?

> In follow-ups where we add in styles that aren't generated inline, we might use a different return type that could include both the generated styles and the classnames that refer to those styles. But we can play with all that in follow-ups, I just wanted to jot down the thought 😀

Good point. Ultimately I guess we should return some sort of object with an `output` key or something.

Reply:
> Would it still be better to move them given that?

I think it's perfectly fine to keep it where it is for now, I was mostly just thinking out loud for what we might explore in follow-ups. I'll switch this conversation to resolved!